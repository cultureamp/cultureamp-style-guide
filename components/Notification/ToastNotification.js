// @flow
import * as React from 'react';
import GenericNotification from './GenericNotification';
import type { NotificationType } from './GenericNotification';

type Props = {|
  type: NotificationType,
  title: string,
  children: React.Node,
  autohide?: false | true | 'hideCloseIcon',
  onHide?: () => void,
|};

const ToastNotification = (props: Props) => {
  const persistent = props.autohide === 'hideCloseIcon';
  return (
    <GenericNotification
      style="toast"
      persistent={persistent}
      type={props.type}
      title={props.title}
      autohide={Boolean(props.autohide)}
      onHide={props.onHide}
    >
      {props.children}
    </GenericNotification>
  );
};

export default ToastNotification;
