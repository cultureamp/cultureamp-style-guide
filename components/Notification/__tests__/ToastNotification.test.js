import React from 'react';
import { mount } from 'enzyme';
import sinon from 'sinon';
import ToastNotification from '../ToastNotification';

test('The basic notification renders correctly', () => {
  const notification = mount(
    <ToastNotification type="positive" title="Success!">
      It worked!
    </ToastNotification>
  );
  expect(notification.find('.cancel')).toHaveLength(1);
  expect(notification).toMatchSnapshot();
});

test('You can hide the close icon on an autohide toast notification', () => {
  const notification = mount(
    <ToastNotification type="negative" title="Info" autohide hideCloseIcon>
      Currently processing...
    </ToastNotification>
  );
  expect(notification.find('.cancel')).toHaveLength(0);
  expect(notification).toMatchSnapshot();
});

test('Autohide works as expected', () => {
  jest.useFakeTimers();

  const onHide = sinon.spy();
  const notification = mount(
    <ToastNotification
      type="positive"
      title="Success!"
      onHide={onHide}
      autohide
    >
      This is my positive notification
    </ToastNotification>
  );

  // On the first frame, 'fading' should be present so the notification transitions in from a hidden state
  expect(notification.find('div.notification').hasClass('fading')).toBe(true);

  // The element should then immediately transition in
  jest.advanceTimersByTime(1);
  notification.update();
  expect(notification.find('div.notification').hasClass('fading')).toBe(false);

  // The notification should still be visible for the first few seconds
  jest.advanceTimersByTime(4000);
  notification.update();
  expect(notification.find('div.notification').hasClass('fading')).toBe(false);

  // After 5s, the element should fade out, but the onHide will not trigger yet.
  jest.advanceTimersByTime(1000);
  notification.update();
  expect(notification.find('div.notification').hasClass('fading')).toBe(true);
  expect(onHide.callCount).toBe(0);

  // After the fade out animation has had time to run, the onHide handler should trigger.
  notification.find('div.notification').simulate('transitionend', {
    propertyName: 'margin-top',
  });
  notification.update();
  expect(onHide.callCount).toBe(1);
});
