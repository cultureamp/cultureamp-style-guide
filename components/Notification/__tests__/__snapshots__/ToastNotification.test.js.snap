// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The basic notification renders correctly 1`] = `
<ToastNotification
  autohide={false}
  hideCloseIcon={false}
  title="Success!"
  type="positive"
>
  <GenericNotification
    autohide={false}
    persistent={false}
    style="toast"
    title="Success!"
    type="positive"
  >
    <div
      className="notification positive toast hidden"
      onTransitionEnd={[Function]}
      style={
        Object {
          "marginTop": undefined,
        }
      }
    >
      <div
        className="icon"
      >
        <Icon
          icon={null}
          inheritSize={true}
          role="presentation"
        >
          <svg
            aria-hidden={true}
            className="icon inheritSize"
            focusable="false"
            role="presentation"
            viewBox="viewBox"
          >
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </div>
      <div
        className="textContainer"
      >
        <span
          className="title"
        >
          Success!
        </span>
        <span
          className="textWrapper"
        >
          <span
            className="text"
          >
            It worked!
          </span>
        </span>
      </div>
      <CancelButton
        onClick={[Function]}
      >
        <button
          className="cancel"
          onClick={[Function]}
        >
          <span
            className="cancelInner"
          >
            <Icon
              icon={null}
              inheritSize={true}
              role="img"
              title="close notification"
            >
              <svg
                aria-labelledby="id-2-title"
                className="icon inheritSize"
                focusable="false"
                role="img"
                viewBox="viewBox"
              >
                <title
                  id="id-2-title"
                >
                  close notification
                </title>
                <use
                  xlinkHref="#id"
                />
              </svg>
            </Icon>
          </span>
        </button>
      </CancelButton>
    </div>
  </GenericNotification>
</ToastNotification>
`;

exports[`You cannot hide the close icon unless it is an autohide 1`] = `
<ToastNotification
  autohide={false}
  hideCloseIcon={true}
  title="Info"
  type="negative"
>
  <GenericNotification
    autohide={false}
    persistent={false}
    style="toast"
    title="Info"
    type="negative"
  >
    <div
      className="notification negative toast hidden"
      onTransitionEnd={[Function]}
      style={
        Object {
          "marginTop": undefined,
        }
      }
    >
      <div
        className="icon"
      >
        <Icon
          icon={null}
          inheritSize={true}
          role="presentation"
        >
          <svg
            aria-hidden={true}
            className="icon inheritSize"
            focusable="false"
            role="presentation"
            viewBox="viewBox"
          >
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </div>
      <div
        className="textContainer"
      >
        <span
          className="title"
        >
          Info
        </span>
        <span
          className="textWrapper"
        >
          <span
            className="text"
          >
            Currently processing...
          </span>
        </span>
      </div>
      <CancelButton
        onClick={[Function]}
      >
        <button
          className="cancel"
          onClick={[Function]}
        >
          <span
            className="cancelInner"
          >
            <Icon
              icon={null}
              inheritSize={true}
              role="img"
              title="close notification"
            >
              <svg
                aria-labelledby="id-5-title"
                className="icon inheritSize"
                focusable="false"
                role="img"
                viewBox="viewBox"
              >
                <title
                  id="id-5-title"
                >
                  close notification
                </title>
                <use
                  xlinkHref="#id"
                />
              </svg>
            </Icon>
          </span>
        </button>
      </CancelButton>
    </div>
  </GenericNotification>
</ToastNotification>
`;

exports[`autohide and hideClose icon both work together correctly 1`] = `
<ToastNotification
  autohide={true}
  hideCloseIcon={true}
  title="Info"
  type="negative"
>
  <GenericNotification
    autohide={true}
    persistent={true}
    style="toast"
    title="Info"
    type="negative"
  >
    <div
      className="notification negative toast hidden"
      onTransitionEnd={[Function]}
      style={
        Object {
          "marginTop": undefined,
        }
      }
    >
      <div
        className="icon"
      >
        <Icon
          icon={null}
          inheritSize={true}
          role="presentation"
        >
          <svg
            aria-hidden={true}
            className="icon inheritSize"
            focusable="false"
            role="presentation"
            viewBox="viewBox"
          >
            <use
              xlinkHref="#id"
            />
          </svg>
        </Icon>
      </div>
      <div
        className="textContainer"
      >
        <span
          className="title"
        >
          Info
        </span>
        <span
          className="textWrapper"
        >
          <span
            className="text"
          >
            Currently processing...
          </span>
        </span>
      </div>
    </div>
  </GenericNotification>
</ToastNotification>
`;
